pkg/runner/local_docker.go

```diff
@@ -172,36 +173,75 @@ func (r *LocalDockerRunner) Healthcheck(ctx context.Context, engine api.Engine,
 	return hh.RunChecks(ctx, fix)
 }
 
-func (c *LocalDockerRunner) updateRunResult(template *runtime.RunParams, result *Result) error {
-	events, err := c.syncClient.FetchAllEvents(template)
+// setupSyncClient sets up the sync client if it is not set up already.
+func (r *LocalDockerRunner) setupSyncClient() error {
+	r.lk.Lock()
+	defer r.lk.Unlock()
+
+	if r.syncClient != nil {
+		return nil
+	}
+
+	err := os.Setenv(ss.EnvServiceHost, "127.0.0.1")
 	if err != nil {
 		return err
 	}
 
-	for _, e := range events {
-		// for now we emit only outcome OK events, so no need for more checks
-		if e.SuccessEvent != nil {
-			se := e.SuccessEvent
-			o := result.Outcomes[se.TestGroupID]
-			o.Ok = o.Ok + 1
-		}
+	r.syncClient, err = ss.NewGenericClient(context.Background(), logging.S())
+	if err != nil {
+		return err
 	}
 
-	result.Outcome = task.OutcomeSuccess
-	if len(result.Outcomes) == 0 {
-		result.Outcome = task.OutcomeFailure
+	return nil
+}
+
+func (r *LocalDockerRunner) collectOutcomes(ctx context.Context, result *Result, tpl *runtime.RunParams) error {
+	eventsCh, err := r.syncClient.SubscribeEvents(ctx, tpl)
+	if err != nil {
+		return err
 	}
-	for g := range result.Outcomes {
-		if result.Outcomes[g].Total != result.Outcomes[g].Ok {
+
+	go func() {
+		running := true
+		for running {
+			select {
+			case <-ctx.Done():
+				running = false
+			case e := <-eventsCh:
+				// for now we emit only outcome OK events, so no need for more checks
+				if e.SuccessEvent != nil {
+					se := e.SuccessEvent
+					o := result.Outcomes[se.TestGroupID]
+					o.Ok = o.Ok + 1
+				}
+			}
+		}
+
+		result.Outcome = task.OutcomeSuccess
+		if len(result.Outcomes) == 0 {
 			result.Outcome = task.OutcomeFailure
-			break
 		}
-	}
+
+		for g := range result.Outcomes {
+			if result.Outcomes[g].Total != result.Outcomes[g].Ok {
+				result.Outcome = task.OutcomeFailure
+				break
+			}
+		}
```

Comment:
@hacdias same comment here - I think we need a channel/notification to the runner when this is done, so that we return only after `result` has been updated, no?

Reply:
Fixed on c5fd5d3.